src/fileutil.c

```diff
@@ -15,29 +15,94 @@
  */
 
 #include <defines.h>
+#include <errno.h>
+#include <pwd.h>                /* for getpwnam_r() */
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
 
 extern char* scenario_path;
 
+int expand_user_path(const char* path, char* expanded_home_path /*The buffer*/, size_t buflen)
+{
+    if (path[0] != '~') {                       /* We have nothing to expand here */
+        return 1;
+    }
+
+    memset(expanded_home_path, '\0', buflen);
+    char* home_dir = NULL;
+
+    if (path[1] == '\0' || path[1] == '/') {                                                    /* '~/path' case */
+        home_dir = getenv("HOME");
+        if (home_dir == NULL) {
+            home_dir = getenv("USERPROFILE");
+        } else {
+            snprintf(expanded_home_path, buflen - 1, "%s%s", home_dir, path + 1);
+        }
+    } else {
+        const char* first_slash = NULL;
+        first_slash = strchr(path, '/');                                                        /* substring starting from '/' */
+        size_t linux_username_limit = 32;                                                       /* As of now */
+        char* username = NULL;
+        if ((first_slash != NULL) && ((first_slash - (path + 1)) <= linux_username_limit)) {    /* '~someuser/blah' case */
+            username = strndup(path + 1, first_slash - (path + 1));
+        } else {                                                                                /* '~someuser' case, there is no file, just username */
+            return -1;
+        }
+
+        struct passwd pwd;
+        struct passwd* result;
+        size_t bufsize = sysconf(_SC_GETPW_R_SIZE_MAX);
+        char buffer[bufsize + 1];
+        int retcode = getpwnam_r(username, &pwd, buffer, bufsize, &result);
```

Comment:
You can `free(username)` right after this line, right?

Reply:
Done, Thanks! will push all changes at once